### PR TITLE
Decompose splines with t flexibly updated

### PR DESCRIPTION
**Pull request by weihsinyeh**
The evaluation testbed is [font-edit](https://github.com/sysprog21/mado/tree/main/tools/font-edit), where I modified the implementation to use fixed-point arithmetic.
points_comparison
dcj_calls_comparison

The font generated by the [original implementation](https://github.com/sysprog21/mado/commit/6d85b4280db3288485c9261b512ec1640697a678) is as follows:
original_output

The font generated by the implementation of decomposing splines with flexibly updated t is as follows:
my_way_output

### CODE REVIEW

File: src/spline.c [hunk 1]
```diff
@@ -93,14 +93,37 @@ static void _twin_spline_decompose(twin_path_t *path,
...
+            if (best == t)
+                break;
+
+            _de_casteljau(spline, t, &left, &right);
+
+            error = _twin_spline_error_squared(&left);
+            if (error < best_error)
+                break;
+
+            /* Left is close enough to the bezier curve */
```
Comment (jserv) [inline src/spline.c:112]: Capitalize "Bézier" curve.

File: src/spline.c [hunk 2]
```diff
@@ -93,14 +93,37 @@ static void _twin_spline_decompose(twin_path_t *path,
...
+            if (error <= tolerance_squared) {
+                best_error = error;
+                best = t;
+                /* Try to find better t */
+                lo = t;
+            } else {
+                /* Find good t already and cannot find better t */
```
Comment (jserv) [inline src/spline.c:119]: Improve the comment.

**Comment by jserv** [comment]
Replace "See also: https://github.com/sysprog21/mado/issues/2" with "Close #2" in the git commit message, so that this issue can be closed automatically once this pull request is merged.
See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

File: src/spline.c (hunk 1 above)
Comment (jouae) [reply src/spline.c:112]: @weihsinyeh 
Any formula, equation, or theorem related to a person's name should have the name capitalized. For example, Euler formula, Lagrange theorem, Newton's method etc.

This is a sign of respect and basic courtesy towards these great individuals.

I have also been reminded of this many times.

File: src/spline.c [hunk 3]
```diff
@@ -93,14 +93,37 @@ static void _twin_spline_decompose(twin_path_t *path,
...
+
+            _de_casteljau(spline, t, &left, &right);
+
+            error = _twin_spline_error_squared(&left);
+            if (error < best_error)
+                break;
+
+            /* Left is close enough to the bezier curve */
+            if (error <= tolerance_squared) {
+                best_error = error;
```
Comment (jouae) [inline src/spline.c:114]: I suggest renaming `best_error` to either `farthest_distance` or `optimal_distance`, as your intent isn't clearly reflected in the code.

Based on the current variable name, the purpose of `best_error` isn't immediately obvious.

Initially, I thought it represented a lower bound, but after reviewing your development process, I realized its actual purpose is to find the farthest distance from point a to point d.

Also, rename `best` to `best_t` or `optimal_t`. 
Avoid using only adjectives as variable names. Make sure what the variable it is.

Refs:
1. https://hackmd.io/UIOMextwQbqIcZaKuFkl4Q#iterative-splines

**Review by jouae — COMMENTED** [review]
>Although this approach requires more computation to decompose the
spline into fewer segments, the extra effort is justified for improved
fit.

You can include statistical data such as mean, standard deviation, etc., in the PR and present it simply in the commit. Remember to mention the number of observations used.

**Comment by jouae** [comment]
* Left side one is the origin and right side one is your method
left_origin_right_modified


I merged two GIFs and slowed down the switching speed so that more people can clearly see your contributions.

File: src/spline.c (hunk 3 above)
Comment (jserv) [reply src/spline.c:114]: > Also, rename best to `best_t` or `optimal_t`.

The suffix `_t` refers to the derived type using `typedef`. In this case, we can say, `t_optimal`.

**Comment by weihsinyeh** [comment]
original - Average iterations per point: 1.99
original - Average points per character: 18.89
flexibly update - Average iterations per point: 4.53
flexibly update - Average points per character: 16.30

**Comment by jserv** [comment]
Refine the git commit messages upon your experiments.

**Review by jouae — APPROVED** [review]

**Comment by jserv** [comment]
FYI: [vector-graphics/src/bezier.h](https://github.com/rafaellg79/vector-graphics/blob/master/src/bezier.h) is an interesting implementation.

**Review by jserv — CHANGES_REQUESTED** [review]
Use `git rebase` rather than `git merge`.

File: src/spline.c [hunk 4]
```diff
@@ -11,38 +11,38 @@ typedef struct _twin_spline {
...
  */
 static void _lerp(twin_spoint_t *a,
                   twin_spoint_t *b,
-                  int shift,
+                  twin_dfixed_t t,
                   twin_spoint_t *result)
 {
-    result->x = a->x + ((b->x - a->x) >> shift);
-    result->y = a->y + ((b->y - a->y) >> shift);
+    result->x = a->x + (((twin_dfixed_t) (b->x - a->x) * t) >> 8);
```
Comment (jserv) [inline src/spline.c:23]: Can you avoid hardcoded `>> 8`? Use the helper macro instead.

**Comment by jserv** [comment]
Google's skia performs subdivision using de Casteljau's algorithm, that is, repeated linear interpolation between adjacent points. See [src/base/SkBezierCurves.cpp](https://github.com/google/skia/blob/main/src/base/SkBezierCurves.cpp).

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase the latest `main` branch.

**Review by jserv — CHANGES_REQUESTED** [review]
Since refining and reviewing this pull request may take some time, I suggest separating the "Fix the typo" commit into a new pull request.

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase the latest `main` branch.

**Comment by jserv** [comment]
> Take the right half of the spline as another standard when finding the optimal $t$. Because if the right half of the spline is also within the tolerance, the right half of the spline will be the last curve segment. Thus, in this case, there is no need to find better $t$, which aims to prevent more line segments from being generated in the future.

Could you assess the appropriate values for tolerance?

**Comment by weihsinyeh** [comment]
The following is the graph to illustrate the first time of calling the function De Casteljau's is redundant calculation.
renderspline drawio (3)
I observed that when we apply shift 0.25 in the flexible setting, it will increase the average points per character. This indicates that when we limit the scope of 't' to [0, 0.25], we will fail to find some optimal point that has maximum curvature in the scope of 't' to [0.25, 0.5]. To verify it, we limit the scope of 't' to just [0,0.25]. The experiment is as follows:
```
Original (shift2) - Average number of _de_casteljau calls per point: 1.51
Original (shift2) - Average points per character: 18.98
Original (shift2) (limit scope) - Average number of _de_casteljau calls per point: 1.18
Original (shift2) (limit scope) - Average points per character: 22.57
```
Therefore, limiting the scope can decrease the average number of function '_de_casteljau' calls per point and increase the number of points. This means that these points do not always have the maximum curvature. In other words, some optimal point is located when the scope of t is [0.25, 1]. Therefore, in this pull request, we can dynamically increase the shift value without limitting the scope.

image

**Comment by weihsinyeh** [comment]
Another Idea:
There may be a new approach to rendering Bézier curves. Instead of calculating the points 'AB' 'BC' 'ABBC' in the left segment, which are colored red, we could calculate positions 'D' 'CD' 'BCCD' using only the right segment, which are colored blue.
renderspline drawio (6)

This is because after rendering this turn, we only use the outcome of the right (blue) of the original spline to render an unhandled segment of spline. The left of the original spline (red) is dropped. What's even worse, not only the left of the original spline will be dropped, but also the intermediate results will be dropped (black font).

**Comment by weihsinyeh** [comment]
Assess the appropriate values for tolerance when rendering B´ezier Curve.

image

The following images show the effect of four settings on 56 fonts, which are rendered using curves. Only these fonts are used to calculate the average number of _de_casteljau calls per point and the average points per character.

Original (tolerance 0.0625) : 
all_font_original

Original (tolerance 0.25) :
all_font_original_tolerance_0 25

Original (shift2) : 
all_font_original_shift2

Original (shift2) (tolerance 0.25) :
all_font_original_shift2_tolerance_0 25

**Comment by weihsinyeh** [comment]
Each time a spline is rendered, multiple iterations are performed. In each iteration, there are multiple  _de_casteljau function calls, which aims to find the most suitable shift value. The y-axis represents the number of shift attempts within one iteration. The x-axis represents the iterations of one spline. Since each spline has a different number of iterations, the iteration count is normalized to the range (0, 1). This allows us to observe the variation in the number of shift attempts at different stages of the iterations during the process of rendering a spline.
image
From the distribution above, we can see that in the original method, it takes an average of more than two shifts to find the optimal value during the first 60% of the spline rendering. It is also evident that as the rendering of a spline process reaches the later stages, the number of shifts gradually decreases. Therefore, based on the first finding, we set the initial shift to 2. Additionally, due to its gradually decreasing nature, the shifts used in different iterations of each spline will be stored as a global variable to record the value of the last shift change. This way, in the next iteration, we can directly use this value instead of starting from a shift of 2 again.

**Review by jserv — CHANGES_REQUESTED** [review]
Refine the git commit messages to reflect the latest progress.

File: src/spline.c [hunk 5]
```diff
@@ -91,16 +92,19 @@ static void _twin_spline_decompose(twin_path_t *path,
...
-        do {
+        while (true) {
             _de_casteljau(spline, shift, &left, &right);
-            shift++;
-        } while (!is_flat(&left, tolerance_squared));
+            if (is_flat(&left, tolerance_squared)) {
+                if (shift > 1)
+                    shift--;
+                break;
+            } else
```
Comment (jserv) [inline src/spline.c:105]: Replace `} else` with simplified `}`.

**Review by jserv — CHANGES_REQUESTED** [review]
Squash the git commits. A single and informative one is preferable.

**Comment by weihsinyeh** [comment]
Assess the appropriate values for tolerance when rendering B´ezier Curve.
image

File: src/spline.c [hunk 6]
```diff
@@ -91,16 +92,19 @@ static void _twin_spline_decompose(twin_path_t *path,
 {
     /* Draw starting point */
     _twin_path_sdraw(path, spline->a.x, spline->a.y);
-
+    int shift = 2;
```
Comment (jserv) [inline src/spline.c:95]: Add comments.

**Pull request merged**